python/graphstorm/graphpartitioning/metadata_schema.py

```diff
@@ -0,0 +1,706 @@
+        etype_counts = self._data[STR_NUM_EDGES_PER_TYPE]
+        for count in etype_counts:
+            assert (
+                isinstance(count, int) and count > 0
+            ), "Invalid value for No. of nodes per type"
+
+        prefix_sum = np.cumsum([0] + etype_counts)
+        starts = prefix_sum[:-1]
+        ends = prefix_sum[1:]
+        ranges = zip(starts, ends)
+
+        self._global_eid_offsets = dict(zip(etypes, ranges))
+
+    def _init_global_nid_offsets(self):
+        """Function to initialize the global_node_id offsets for all the
+        node types present in the input graph.
+
+        global_node_ids, or global_nids, are unique ids assigned to each
+        node of the input graph. These ids are expected to be unique ids
+        across the input graph.
+
+        The range of global_node_ids for each node type in the graph are
+        stored in a dictionary where key-value pairs are ntypes and a tuple
+        which indicates the starting and ending value (exclusive) of the
+        global_node_ids. For instance,
+            ntype1: [a, b) are the key-value pairs.
+            where ntype1 is a string describing the node type
+            and [a, b) is the range of global_node_ids for the node type
+            `ntype1`.
+        """
+        ntypes = self._data[STR_NODE_TYPE]
+        ntype_counts = self._data[STR_NUM_NODES_PER_TYPE]
+        for count in ntype_counts:
+            assert (
+                isinstance(count, int) and count > 0
+            ), "Invalid value for No. of nodes per type"
+
+        prefix_sum = np.cumsum([0] + ntype_counts)
+        starts = prefix_sum[:-1]
+        ends = prefix_sum[1:]
+        ranges = zip(starts, ends)
+        self._global_nid_offsets = dict(zip(ntypes, ranges))
+
+    def _init_etype_maps(self):
+        """Initialization function to build unique id maps for edge types of
+        the input graph. This function creates a list of edge types present in
+        the graph, and unique-id <-> edge type and reverse maps.
+        """
+        self._etypes = self._data[STR_EDGE_TYPE]
+        self._etype_id_map = {
+            etype_name: idx for idx, etype_name in enumerate(self._etypes)
+        }
+        self._id_etype_map = dict(enumerate(self._etypes))
+
+    def _init_etype_features(self):
+        """Initialization function to build a map in which key are edge types
+        in the graph and values are a list of features for the corresponding
+        edge type, if present. For instance,
+            etype1: [feature1, feature2, ....] where
+            etype1 is a edge type from the graph and
+            [feature1, feature2, ...] is a list of features associated with
+            the etype1 edge type.
+        """
+        self._etype_features = {}
+        edge_data = self._data.get(STR_EDGE_DATA, {})
+        for etype in self._data[STR_EDGE_TYPE]:
+            features = edge_data.get(etype, {})
+            if len(features) > 0:
```

Comment:
this is not a normal way of checking if `etype` exists. this is a common problem. i see you are doing it everywhere. i pointed out one place, you should change all places.